File: com/dev/leetcodehard/EditDistance.java
/*
https://leetcode.com/problems/edit-distance/

status - completed
 */

package com.dev.leetcodehard;

import com.dev.logger.basePrinter;

class EditDistance extends basePrinter{

    static int minDistance(String word1, String word2){
        int m = word1.length(), n = word2.length();
        int[][] dp = new int[m + 1][n + 1];

        for (int i = 0; i <= m; i++){
            dp[i][0] = i;
        }
        for (int j = 0; j <= n; j++){
            dp[0][j] = j;
        }

        for (int i = 1; i <= m; i++){
            for (int j = 1; j <= n; j++){
                if (word1.charAt(i - 1) == word2.charAt(j - 1)){
                    dp[i][j] = dp[i - 1][j - 1];
                }
                else{
                    int replace = dp[i - 1][j - 1];
                    int delete = dp[i - 1][j];
                    int insert = dp[i][j - 1];
                    dp[i][j] = 1 + Math.min(replace, Math.min(delete, insert));
                }
            }
        }

        return dp[m][n];
    }

    public static void main(String[] args) {
        logp("input word1=horse word2=ros expected 3 actual " + minDistance("horse", "ros"));
        logp("input word1=intention word2=execution expected 5 actual " + minDistance("intention", "execution"));
        logp("input word1= word2=abc expected 3 actual " + minDistance("", "abc"));
    }
}

File: python/leetcodehard/edit_distance.py
"""
https://leetcode.com/problems/edit-distance/

status - completed
"""

from base_logger.logging_event import create_logger


class Solution:

    log = create_logger(__name__)

    def minDistance(self, word1, word2):
        rows = len(word1)
        cols = len(word2)
        dp = [[0] * (cols + 1) for _ in range(rows + 1)]

        for i in range(rows + 1):
            dp[i][0] = i
        for j in range(cols + 1):
            dp[0][j] = j

        for i in range(1, rows + 1):
            for j in range(1, cols + 1):
                if word1[i - 1] == word2[j - 1]:
                    dp[i][j] = dp[i - 1][j - 1]
                else:
                    dp[i][j] = 1 + min(dp[i - 1][j - 1], dp[i - 1][j], dp[i][j - 1])

        return dp[rows][cols]


if __name__ == "__main__":
    test_input_1 = ("horse", "ros")
    output = Solution().minDistance(*test_input_1)
    Solution().log.info("Output is %s", output)

    test_input_2 = ("intention", "execution")
    output = Solution().minDistance(*test_input_2)
    Solution().log.info("Output is %s", output)
